src/mapping/LinearCellInterpolationMapping.cpp

```diff
@@ -0,0 +1,116 @@
+
+namespace mapping {
+
+LinearCellInterpolationMapping::LinearCellInterpolationMapping(
+    Constraint constraint,
+    int        dimensions)
+    : BarycentricBaseMapping(constraint, dimensions)
+{
+  PRECICE_CHECK(getDimensions() == 2, "Volume mapping not available in 3D.");
+  if (constraint == CONSISTENT) {
+    setInputRequirement(Mapping::MeshRequirement::FULL);
+    setOutputRequirement(Mapping::MeshRequirement::VERTEX);
+  } else if (constraint == CONSERVATIVE) {
+    setInputRequirement(Mapping::MeshRequirement::VERTEX);
+    setOutputRequirement(Mapping::MeshRequirement::FULL);
+  } else {
+    PRECICE_ASSERT(constraint == SCALEDCONSISTENT, constraint);
+    setInputRequirement(Mapping::MeshRequirement::FULL);
+    setOutputRequirement(Mapping::MeshRequirement::FULL);
+  }
+
+  PRECICE_CHECK(constraint != SCALEDCONSISTENT, "Volume mapping doesn't support scaled-consistent mappings.");
+}
+
+void LinearCellInterpolationMapping::computeMapping()
+{
+  PRECICE_TRACE(input()->vertices().size(), output()->vertices().size());
+  const std::string     baseEvent = "map.vci.computeMapping.From" + input()->getName() + "To" + output()->getName();
+  precice::utils::Event e(baseEvent, precice::syncMode);
+  PRECICE_ASSERT(getDimensions() == 2, "Volume mapping not available in 3D.");
+
+  // Setup Direction of Mapping
+  mesh::PtrMesh origins, searchSpace;
+  if (hasConstraint(CONSERVATIVE)) {
+    PRECICE_DEBUG("Compute conservative mapping");
+    origins     = input();
+    searchSpace = output();
+  } else {
+    PRECICE_DEBUG("Compute consistent mapping");
+    origins     = output();
+    searchSpace = input();
+  }
+
+  const auto &fVertices = origins->vertices();
+
+  if (getDimensions() == 2) {
+    if (!fVertices.empty() && searchSpace->triangles().empty()) {
+      PRECICE_WARN("2D Mesh \"{}\" does not contain triangles. "
+                   "Linear cell interpolation falls back to nearest projection mapping.",
+                   searchSpace->getName());
+    }
+  } else {
+    // TODO, shouldn't be reached
+    PRECICE_ERROR("Linear Cell interpolation in 3D not implemented yet. This line should be unreachable");
+  }
+
+  // Amount of nearest elements to fetch for detailed comparison.
+  // This safety margin results in a candidate set which forms the base for the
+  // local nearest projection and counters the loss of detail due to bounding box generation.
+  // @TODO Add a configuration option for this factor
+  constexpr int nnearest = 4;
+
+  auto &                                 index = searchSpace->index();
+  utils::statistics::DistanceAccumulator fallbackStatistics;
+
+  _interpolations.clear();
+  _interpolations.reserve(fVertices.size());
+
+  for (const auto &fVertex : fVertices) {
+    // Find tetrahedra (3D) or triangle (2D) or fall-back on NP
+    auto match = index.findCellInterpolation(fVertex.getCoords(), nnearest);
+    _interpolations.push_back(std::move(match.polation));
+    if (match.distance != 0.0) {
```

Comment:
We need to use a tolerance-based check here: `math::equals(match.distance, 0.0)`

Reply:
Are you sure ? An "almost zero" means there has been some projection, as findCelInterpolation returns an exact 0 if a triangle was used.

Reply:
Not necessarily. The tolerance check covers rounding errors etc.